### PR TITLE
feat: restricting days date validation

### PR DESCRIPTION
**Pull request by wenhaogoh**
## Problem

- Currently, the date validation only allows for 3 types of validations: Disallow past dates, Disallow future dates and Custom Date Range. However, form users have been reaching out for providing more customization in the date field so that they can use formsg to cater to their specific needs. Hence, to cater to such customisation of date validation, there needs to be additional types of date validations added to the existing ones.
- Design document [here](https://docs.google.com/document/d/13G0GQ0cpLgC5r3rOAVKlzTPaFNi-RkD2sOGHRcsTOaY/edit?usp=sharing)

Closes FRM-432

## Solution

- Added `Customise availability during the week` restriction

## Screenshots

- Error messages
  - None selected
Screenshot 2022-10-13 at 1 07 12 AM
  - No available dates
    Screenshot 2022-10-19 at 2 09 02 PM
  - User manually inputs an invalid date
Screenshot 2022-10-13 at 1 08 37 AM

### CODE REVIEW

**Review by karrui — COMMENTED** [review]
lgtm, but lets wait for react hotfixes to go out before we merge this in.

File: frontend/src/templates/Field/Date/DateField.stories.tsx [hunk 1]
```diff
@@ -164,3 +169,13 @@ ValidationCustomRange.args = {
...
 }
+
+export const ValidationInvalidDaysOfTheWeek = Template.bind({})
+ValidationInvalidDaysOfTheWeek.args = {
+  schema: {
+    ...baseSchema,
+    description: 'Wednesdays and Thursdays are invalid days',
+    invalidDays: [InvalidDaysOptions.Wednesday, InvalidDaysOptions.Thursday],
+  },
+  defaultValue: '2022-07-27',
```
Comment (karrui) [inline frontend/src/templates/Field/Date/DateField.stories.tsx]: might need to update this value to dd/mm/yyyy format

File: shared/utils/date-validation.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  [InvalidDaysOptions.Saturday]: 6,
+  [InvalidDaysOptions.Sunday]: 7,
+}
+
+/**
+ * Convert the days of the week in the invalidDays array
+ * to a number array representing the number representation
+ * of the corresponding day of the week
+ */
+export const convertInvalidDaysOfTheWeekToNumberSet = (
```
Comment (karrui) [inline shared/utils/date-validation.ts:19]: should we add unit tests for the functions in this util file?

Comment (wenhaogoh) [reply shared/utils/date-validation.ts:19]: added!

File: frontend/src/templates/Field/Date/DateField.stories.tsx (hunk 1 above)
Comment (wenhaogoh) [reply frontend/src/templates/Field/Date/DateField.stories.tsx]: fixed!

Comment (LinHuiqing) [reply frontend/src/templates/Field/Date/DateField.stories.tsx]: for reference, fixed in [878610c](https://github.com/opengovsg/FormSG/pull/5136/commits/878610ca649a6bc555ffca1bc573c270b5cb0ff1)

File: shared/utils/date-validation.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  start: Date,
+  end: Date,
+  invalidDays: InvalidDaysOptions[],
+): boolean => {
+  if (invalidDays.length < 7 && differenceInCalendarDays(start, end) >= 6) {
+    return true
+  }
+
+  for (
+    let date = new Date(start);
```
Comment (LinHuiqing) [inline shared/utils/date-validation.ts]:
```suggestion
    let date = start;
```
`start` is already of `Date` type

File: shared/utils/date-validation.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    date <= end;
+    date.setDate(date.getDate() + 1)
+  ) {
+    if (isDateAnInvalidDay(date, invalidDays)) {
+      continue
+    }
+    return true
```
Comment (LinHuiqing) [inline shared/utils/date-validation.ts]:
```suggestion
    if (!isDateAnInvalidDay(date, invalidDays)) {
      return true
    }
```

File: frontend/src/templates/Field/Date/DateField.tsx [hunk 5]
```diff
@@ -39,29 +40,34 @@ export const DateField = ({
...
             loadDateFromNormalizedDate(customMaxDate),
           )
+          break
         }
         default:
-          return false
+          break
       }
+
+      return isDateUnavailable || isDateAnInvalidDay(date, selectedInvalidDays)
```
Comment (LinHuiqing) [inline frontend/src/templates/Field/Date/DateField.tsx:68]: Just curious - Is there a reason why we refactored this to not return immediately when there's an error case?

File: src/app/utils/field-validation/validators/__tests__/date-validation.spec.ts [hunk 6]
```diff
@@ -323,4 +324,60 @@ describe('Date field validation', () => {
...
+      },
+      invalidDays: [InvalidDaysOptions.Wednesday, InvalidDaysOptions.Thursday],
+    })
+    const response = generateNewSingleAnswerResponse(BasicField.Date, {
+      answer: '29 Jul 2022',
+    })
+
+    const validateResult = validateField('formId', formField, response)
+    expect(validateResult.isOk()).toBe(true)
+    expect(validateResult._unsafeUnwrap()).toEqual(true)
```
Comment (LinHuiqing) [inline src/app/utils/field-validation/validators/__tests__/date-validation.spec.ts]: Should check _all_ days of the week which are valid, not just one.

File: src/app/utils/field-validation/validators/__tests__/date-validation.spec.ts [hunk 7]
```diff
@@ -323,4 +324,60 @@ describe('Date field validation', () => {
...
+    })
+    const response = generateNewSingleAnswerResponse(BasicField.Date, {
+      answer: '27 Jul 2022',
+    })
+
+    const validateResult = validateField('formId', formField, response)
+    expect(validateResult.isErr()).toBe(true)
+    expect(validateResult._unsafeUnwrapErr()).toEqual(
+      new ValidateFieldError('Invalid answer submitted'),
+    )
```
Comment (LinHuiqing) [inline src/app/utils/field-validation/validators/__tests__/date-validation.spec.ts]: Should check _all_ days of the week which are invalid, not just one.

File: frontend/src/templates/Field/Date/DateField.tsx (hunk 5 above)
Comment (LinHuiqing) [reply frontend/src/templates/Field/Date/DateField.tsx:68]: Nevermind, going to resolve this as this works too!

File: shared/utils/date-validation.ts (hunk 3 above)
Comment (LinHuiqing) [reply shared/utils/date-validation.ts]: Done in b47dec98

File: shared/utils/date-validation.ts (hunk 4 above)
Comment (LinHuiqing) [reply shared/utils/date-validation.ts]: Done in b47dec98

File: src/app/utils/field-validation/validators/__tests__/date-validation.spec.ts (hunk 6 above)
Comment (LinHuiqing) [reply src/app/utils/field-validation/validators/__tests__/date-validation.spec.ts]: Done in [cadbda4](https://github.com/opengovsg/FormSG/pull/5136/commits/cadbda4009a4db4b46d611dcf6c3d528ef580761)

File: src/app/utils/field-validation/validators/__tests__/date-validation.spec.ts (hunk 7 above)
Comment (LinHuiqing) [reply src/app/utils/field-validation/validators/__tests__/date-validation.spec.ts]: Done in [cadbda4](https://github.com/opengovsg/FormSG/pull/5136/commits/cadbda4009a4db4b46d611dcf6c3d528ef580761)

**Review by LinHuiqing — APPROVED** [review]
LGTM! Thank you my FormSG predecessors @wenhaogoh @siddarth2824 and @karrui for the prior work!

**Pull request merged**